### PR TITLE
EVG-6326: handle renewing Jasper credentials on hosts

### PR DESCRIPTION
**Pull request by Kimchelly**
JIRA: https://jira.mongodb.org/browse/EVG-6326

* Jasper credentials expire when host has been up for 1 year.
* Create job that writes new Jasper credentials to a host and restart Jasper when their credentials are within a week of their expiration.
* If communicating via RPC, attempt to deploy by making requests to using Jasper RPC service. Otherwise, if we fail to deploy credentials with Jasper RPC or the host is communicating over SSH, fall back to using SSH for the credentials deploy.

### CODE REVIEW

File: model/event/host_event.go [hunk 1]
```diff
@@ -110,6 +113,14 @@ func LogHostAgentMonitorDeployFailed(hostId string, err error) {
 	LogHostEvent(hostId, EventHostAgentMonitorDeployFailed, HostEventData{Logs: err.Error()})
 }
 
+func LogHostJasperDeployed(hostId string, revision string) {
+	LogHostEvent(hostId, EventHostJasperDeployed, HostEventData{JasperRevision: revision})
+}
+
+func LogHostJasperDeployFailed(hostId string, err error) {
```
Comment (tychoish) [inline model/event/host_event.go]: should we handle nil errors?

File: model/host/db.go [hunk 2]
```diff
@@ -365,6 +367,46 @@ func FindByFirstProvisioningAttempt() ([]Host, error) {
...
+// credentials will expire within the given cutoff.
+func FindByExpiringJasperCredentials(cutoff time.Duration) ([]Host, error) {
+	deadline := time.Now().Add(cutoff)
+	bootstrapKey := bsonutil.GetDottedKeyName(DistroKey, distro.BootstrapMethodKey)
+	credentialsKey := credentials.Collection
+	expirationKey := bsonutil.GetDottedKeyName(credentialsKey, credentials.TTLKey)
+
+	var hosts []Host
+
+	pipeline := []bson.M{
```
Comment (tychoish) [inline model/host/db.go]: sholdn't we start with a filter of some kind?

File: units/jasper_deploy.go [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+
+	if j.canRetryDeploy() {
+		return errors.Wrap(j.requeueDeployThroughJasper(ctx, false), "could not requeue job without deploy through Jasper")
+	}
+
+	return errors.New("no more Jasper deploy attempts remaining")
+}
+
+func (j *jasperDeployJob) requeueDeployThroughJasper(ctx context.Context, deployThroughJasper bool) error {
+	job := NewJasperDeployJob(j.env, *j.host, j.CredentialsExpiration, deployThroughJasper, fmt.Sprintf("attempt-%d", j.host.JasperDeployAttempts))
```
Comment (tychoish) [inline units/jasper_deploy.go]: is there a way to make the constructor take a pointer to avoid the copies?

File: units/jasper_deploy.go [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+	job.UpdateTimeInfo(amboy.JobTimeInfo{
+		WaitUntil: time.Now().Add(time.Minute),
```
Comment (tychoish) [inline units/jasper_deploy.go]: nice!

File: units/jasper_deploy.go [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+		serviceID := client.ID()
+
+		writeCredentialsOpts := &jasper.CreateOptions{
+			Args: []string{
+				"bash", "-c", writeCredentialsCmd,
+			},
+		}
+		var output string
+		if output, err = j.host.RunJasperProcess(ctx, j.env, writeCredentialsOpts); err != nil {
+			grip.Error(message.Fields{
```
Comment (tychoish) [inline units/jasper_deploy.go]: should we wrap the error?

File: units/jasper_deploy.go [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+			return
+		}
+		// We use this ID to later verify the current running Jasper service.
+		// When Jasper is redeployed, its ID should be different to indicate it
+		// is a new Jasper service.
+		serviceID := client.ID()
+
+		writeCredentialsOpts := &jasper.CreateOptions{
+			Args: []string{
+				"bash", "-c", writeCredentialsCmd,
```
Comment (tychoish) [inline units/jasper_deploy.go]: why do we need to run this in batch? 

how do we ensure that the string is properly quoted?

File: units/jasper_deploy.go [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+
+	writeCredentialsCmd, err := j.host.WriteJasperCredentialsFileCommand(creds)
+	if err != nil {
+		grip.Error(message.Fields{
+			"message": "could not build command to write Jasper credentials file",
+			"host":    j.host.Id,
+			"distro":  j.host.Distro.Id,
+			"job":     j.ID(),
+		})
+		grip.Error(err)
```
Comment (tychoish) [inline units/jasper_deploy.go]: we should combine these errors?

File: units/jasper_deploy.go [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+			grip.Error(message.WrapError(err, message.Fields{
+				"message":   "new service ID should be non-empty",
+				"host":      j.host.Id,
+				"distro":    j.host.Distro.Id,
+				"jasper_id": serviceID,
+				"job":       j.ID(),
+			}))
+			j.AddError(err)
+			return
+		} else if newServiceID == serviceID {
```
Comment (tychoish) [inline units/jasper_deploy.go]: doesn't need the else?

File: units/jasper_deploy.go [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+		"message": "deployed Jasper service with new credentials",
+		"host":    j.host.Id,
+		"distro":  j.host.Distro.Id,
+		"version": settings.HostJasper.Version,
+	})
+	event.LogHostJasperDeployed(j.host.Id, settings.HostJasper.Version)
+
+	// If job succeeded, reset jasper deploy count.
+	if err := j.host.ResetJasperDeployAttempts(); err != nil {
+		grip.Error(message.Fields{
```
Comment (tychoish) [inline units/jasper_deploy.go]: what about this error?

File: units/jasper_deploy.go [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+					"expires":  j.CredentialsExpiration,
+					"attempts": j.host.JasperDeployAttempts,
+					"job":      j.ID(),
+				}))
+				return
+			}
+		}
+	}()
+
+	settings := j.env.Settings()
```
Comment (tychoish) [inline units/jasper_deploy.go]: should this go in the populate if unset

**Comment by tychoish** [comment]
I think there are still a few outstanding comments, and I might have missed the place where you call the Populate function

File: units/jasper_deploy.go (hunk 6 above)
Comment (tychoish) [reply units/jasper_deploy.go]: I think you can specify an io.Rearder to the stdin of a processs for execing, which should work?

Comment (tychoish) [reply units/jasper_deploy.go]: redirect might be harder to fake out.

Comment (Kimchelly) [reply units/jasper_deploy.go]: The bytes of the io.Reader won't be serialized when sending the CreateOptions over RPC/SSH.

Comment (Kimchelly) [reply units/jasper_deploy.go]: I made [a ticket](https://jira.mongodb.org/browse/MAKE-865) to handle standard input for remote processes.

**Pull request merged**